airflow/upgrade/rules/airflow_macro_plugin_removed.py

```diff
@@ -39,9 +39,16 @@ def _check_file(self, file_path):
         problems = []
         class_name_to_check = self.MACRO_PLUGIN_CLASS.split(".")[-1]
         with open(file_path, "r") as file_pointer:
-            for line_number, line in enumerate(file_pointer, 1):
-                if class_name_to_check in line:
-                    problems.append(self._change_info(file_path, line_number))
+            try:
+                for line_number, line in enumerate(file_pointer, 1):
+                    if class_name_to_check in line:
+                        problems.append(self._change_info(file_path, line_number))
+            except UnicodeDecodeError:
+                raise Exception(
+                    "Unable to read file {}, it may need to be added to the .airflowignore".format(
```

Comment:
I think it would be much better to fix the problem as well for non-default encodings.

As of python 3.4 there is this function in the importlib:

https://docs.python.org/3/library/importlib.html?highlight=importlib#importlib.util.decode_source

And it will read the source code of python, respecting the encoding defined in the source code( the `# coding = ` comment). This way it will be more universal and handle the cases where non-utf8 encoding is used (which might be important in many countries). 

I am not sure if similar approach is used in other rules - if so, the we can make a common approach and correct it in all those places.

Reply:
I think this might run into the same [python 2.7](https://docs.python.org/2/library/importlib.html) issue as the [other comment you made](https://github.com/apache/airflow/pull/13371/files#r550270606) 😮  I'm not opposed to changing this logic, if you have an alternate suggestion! I think the issue I was running into had to deal with the fact that the file was binary, not that it was in a different encoding (as I imagine `open` would handle other string encodings appropriately...but admittedly I know less about that side of things).

Reply:
This is no problem. Just skip the test for non-py2.7. The upgradecheck is supposed to be run after people migrate to Python 3.6+. Something like that (or using six) should do.

```
@pytest.mark.skipif(sys.version_info < (3, 6), reason="requires python3.6 or higher")
```

Reply:
I think we should be completely safe to put skipIf < 3.6 on all upgrade check tests @turbaszek @dimberman ?

Reply:
Gotcha! This was caught by the `Compile code using python2` pre-commit hook (run via breeze), which is what clued me in initially.

Reply:
Ah ok. Then if it a that stage, that might be a bit more complex, but just a bit. Simply exclude upgrade_check tests + upgrade_check sources in those two pre-commit checks (they already contain some exclusions):  https://github.com/apache/airflow/blob/748d05f519c0e5427d2a46763591f03697c4d585/.pre-commit-config.yaml#L266

Reply:
Can do! I'll wait until some others chime in on whether or not this can be 3.6+ exclusively, then get to work on that if we get the 👍  🙂

Reply:
Just want to check back in on this @potiuk @turbaszek @dimberman - we can make the upgrade check Python 3.6+ (or even just the tests Py3.6+ if we'd like), but currently the package still supports 2.7 (https://github.com/apache/airflow/blob/v1-10-stable/airflow/upgrade/setup.cfg#L35). Should we bump up these requirements to 3.6? I agree that it makes sense to do so, since this is a utility for upgrading to Airflow 2.0 which is 3.6+.

Reply:
The `airflow upgrade_check` is intended to run on `1.10.X` version of Airflow so I'm afraid that it has to support Py2.7

Reply:
Thanks @turbaszek! With that in mind @potiuk, I'll make sure this is python 2.7 compatible. I'm going to be changing it up a bit though to raise these issues as problems rather than an exception, per the comments below.